src/daft-logical-plan/src/optimization/rules/reorder_joins/join_graph.rs

```diff
@@ -771,9 +802,20 @@ impl JoinGraphBuilder {
                 // Since we hit a join, we need to process the linear chain of Projects and Filters that were encountered starting
                 // from the plan at the root of the linear chain to the current plan.
                 LogicalPlan::Join(Join {
-                    left_on, join_type, ..
-                }) if *join_type == JoinType::Inner && !left_on.is_empty() => {
-                    self.process_linear_chain(root_plan, plan);
+                    on,
+                    join_type: JoinType::Inner,
+                    ..
+                }) => {
+                    let mut on = on.clone();
+                    let (left_on, _, _) = on.pop_equi_preds();
+
+                    if left_on.is_empty() || on.inner().is_some() {
```

Comment:
Nit: here and elsewhere we do `on.inner().is_some()`. It would be nice to have something like a `on.has_remaining_non_equi_preds()` or something